vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -0,0 +1,907 @@
+  {
+    ldDst->outcome = new CCB::conditional_contexual_bandit_outcome();
+    ldDst->outcome->probabilities = v_init<ACTION_SCORE::action_score>();
+
+    ldDst->outcome->cost = ldSrc->outcome->cost;
+    copy_array(ldDst->outcome->probabilities, ldSrc->outcome->probabilities);
+  }
+
+  copy_array(ldDst->explicit_included_actions, ldSrc->explicit_included_actions);
+  ldDst->type = ldSrc->type;
+}
+
+ACTION_SCORE::action_score convert_to_score(const substring& action_id_str, const substring& probability_str)
+{
+  auto action_id = static_cast<uint32_t>(int_of_substring(action_id_str));
+  auto probability = float_of_substring(probability_str);
+  if (nanpattern(probability))
+    THROW("error NaN probability: " << probability_str);
+
+  if (probability > 1.0)
+  {
+    std::cerr << "invalid probability > 1 specified for an outcome, resetting to 1.\n";
+    probability = 1.0;
+  }
+  if (probability < 0.0)
+  {
+    std::cerr << "invalid probability < 0 specified for an outcome, resetting to 0.\n";
+    probability = .0;
+  }
+
+  return {action_id, probability};
+}
+
+//<action>:<cost>:<probability>,<action>:<probability>,<action>:<probability>,…
+CCB::conditional_contexual_bandit_outcome* parse_outcome(substring& outcome)
+{
+  auto& ccb_outcome = *(new CCB::conditional_contexual_bandit_outcome());
+
+  auto split_commas = v_init<substring>();
+  tokenize(',', outcome, split_commas);
+
+  auto split_colons = v_init<substring>();
+  tokenize(':', split_commas[0], split_colons);
+
+  if (split_colons.size() != 3)
+    THROW("Malformed ccb label");
+
+  ccb_outcome.probabilities = v_init<ACTION_SCORE::action_score>();
+  ccb_outcome.probabilities.push_back(convert_to_score(split_colons[0], split_colons[2]));
+
+  ccb_outcome.cost = float_of_substring(split_colons[1]);
+  if (nanpattern(ccb_outcome.cost))
+    THROW("error NaN cost: " << split_colons[1]);
+
+  split_colons.clear();
+
+  for (size_t i = 1; i < split_commas.size(); i++)
+  {
+    tokenize(':', split_commas[i], split_colons);
+    if (split_colons.size() != 2)
+      THROW("Must be action probability pairs");
+    ccb_outcome.probabilities.push_back(convert_to_score(split_colons[0], split_colons[1]));
+  }
+
+  split_colons.delete_v();
+  split_commas.delete_v();
+
+  return &ccb_outcome;
+}
+
+void parse_explicit_inclusions(CCB::label* ld, v_array<substring>& split_inclusions)
+{
+  for (const auto& inclusion : split_inclusions)
+  {
+    ld->explicit_included_actions.push_back(int_of_substring(inclusion));
+  }
+}
+
+void parse_label(parser* p, shared_data*, void* v, v_array<substring>& words)
+{
+  CCB::label* ld = static_cast<CCB::label*>(v);
+
+  if (words.size() < 2)
+    THROW("ccb labels may not be empty");
+  if (!substring_equal(words[0], "ccb"))
+  {
+    THROW("ccb labels require the first word to be ccb");
+  }
+
+  auto type = words[1];
+  if (substring_equal(type, "shared"))
+  {
+    if (words.size() > 2)
+      THROW("shared labels may not have a cost");
```

Comment:
Should we be throwing or warning?

Reply:
I really do think we should throw here. The data is malformed and we should not continue in an undefined state